pennylane/templates/state_preparations/mottonen.py

```diff
@@ -128,113 +138,101 @@ def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
 
     Returns:
-        scipy.sparse.dok_matrix[np.float64]: a sparse vector representing :math:`\alpha^z_k`
+        array representing :math:`\alpha^{z,k}`
     """
-    alpha_z_k = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
+    indices1 = [
+        [(2 * j - 1) * 2 ** (k - 1) + l - 1 for l in range(1, 2 ** (k - 1) + 1)]
+        for j in range(1, 2 ** (n - k) + 1)
+    ]
+    indices2 = [
+        [(2 * j - 2) * 2 ** (k - 1) + l - 1 for l in range(1, 2 ** (k - 1) + 1)]
+        for j in range(1, 2 ** (n - k) + 1)
+    ]
 
-    for (i, _), om in omega.items():
-        i += 1
-        j = int(np.ceil(i * 2 ** (-k)))
-        s_condition = 2 ** (k - 1) * (2 * j - 1)
-        s_i = 1.0 if i > s_condition else -1.0
-        alpha_z_k[j - 1, 0] = alpha_z_k[j - 1, 0] + s_i * om / 2 ** (k - 1)
+    term1 = np.take(omega, indices=indices1)
+    term2 = np.take(omega, indices=indices2)
+    diff = (term1 - term2) / 2 ** (k - 1)
 
-    return alpha_z_k
+    return np.sum(diff, axis=1)
 
 
 def _get_alpha_y(a, n, k):
-    r"""Computes the rotation angles alpha for the Y rotations.
+    r"""Computes the rotation angles required to implement the uniformly controlled Y rotation
+    applied to the :math:`k`th qubit.
+
+    The :math:`j`-th angle is related to the absolute values, a, of the desired amplitudes via:
+
+    .. math:: \alpha^{y,k}_j = 2 \arcsin \sqrt{ \frac{ \sum_{l=1}^{2^{k-1}} a_{(2j-1)2^{k-1} +l}^2  }{ \sum_{l=1}^{2^{k}} a_{(j-1)2^{k} +l}^2  } }
 
     Args:
-        omega (float): phase of the input
-        n (int): total number of qubits
+        a (array): absolute values of the state to prepare
+        n (int): total number of qubits for the uniformly-controlled rotation
         k (int): index of current qubit
 
     Returns:
-        scipy.sparse.dok_matrix[np.float64]: a sparse vector representing :math:`\alpha^y_k`
+        array representing :math:`\alpha^{y,k}`
     """
+    indices_numerator = [
+        [(2 * (j + 1) - 1) * 2 ** (k - 1) + l for l in range(2 ** (k - 1))]
+        for j in range(2 ** (n - k))
+    ]
+    numerator = np.take(a, indices=indices_numerator)
+    numerator = np.sum(np.abs(numerator) ** 2, axis=1)
 
-    alpha = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-
-    numerator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-    denominator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-
-    for (i, _), e in a.items():
-        j = int(math.ceil((i + 1) / 2 ** k))
-        l = (i + 1) - (2 * j - 1) * 2 ** (k - 1)
-        is_part_numerator = 1 <= l <= 2 ** (k - 1)
-
-        if is_part_numerator:
-            numerator[j - 1, 0] += e * e
-        denominator[j - 1, 0] += e * e
-
-    for (j, _), e in numerator.items():
-        numerator[j, 0] = math.sqrt(e)
-    for (j, _), e in denominator.items():
-        denominator[j, 0] = 1 / math.sqrt(e)
+    indices_denominator = [
+        [j * 2 ** k + l for l in range(0, int(2 ** k))] for j in range(2 ** (n - k))
```

Comment:
```suggestion
        [j * 2 ** k + l for l in range(2 ** k)] for j in range(2 ** (n - k))
```
Doesn't need to be an int I think?